bigtable/client/row_range.cc

```diff
@@ -0,0 +1,446 @@
+ * Given any key 'prefix', the prefix range is given as the right-open interval:
+ *   ['prefix', successor('prefix')).
+ *
+ * When prefix is of the form "\xFF\xFF\xFF...\xFF", the successor is
+ * "" (empty string). In such a case, we set end key to 'prefix';
+ *
+ */
+RowRange RowRange::prefix_range(std::string prefix) {
+  RowRange tmp;
+  std::string next_prefix = Successor(prefix);
+
+  if (next_prefix.empty()) next_prefix = prefix;
+  tmp.range_.set_start_key_closed(std::move(prefix));
+  if (!next_prefix.empty())  // Infinite range for prefix="".
+    tmp.range_.set_end_key_open(std::move(next_prefix));
+  return tmp;
+}
+
+// An alias for right_open().
+RowRange RowRange::range(std::string begin, std::string end) {
+  return right_open(begin, end);
+}
+
+/*
+ * Increment the given key: gives the permutation string (in the key-space)
+ * that immediately follows the 'key'.
+ * Please note that, for a key of the form "\xFF\xFF\xFF...\xFF", the successor
+ * is "" (empty string).
+ */
+std::string RowRange::Successor(std::string const& key) {
+  if (key.empty()) {
+    return "";
+  }
+  std::size_t position = key.find_last_not_of('\xFF');
+
+  if (position == std::string::npos) {
+    return "";
+  }
+  const char last_char = key[position] + 1;
+  return (key.substr(0, position) + last_char);
+}
+
+/**
+ * A range  can be of folling types:
+ *  - start is NOT_SET or set to "",  and end is NOT_SET: infinite range
+ *    spanning the complete domain of the row key.
+ *  - start key set to a non-empty 'x', and end is NOT_SET: one-sided infinite
+ *    range starting from x.
+ */
+bool RowRange::contains(std::string_view row) const {
+  KeyType start_type{GetStartKeyType()}, end_type{GetEndKeyType()};
+  std::string start_key{GetStartKey()}, end_key{GetEndKey()};
+  bool result = true;
+
+  if (row.empty()) return true;
+
+  // Check for start key.
+  switch (start_type) {
+    case KeyType::OPEN:
+      result = start_key < row;
+      break;
+    case KeyType::CLOSED:
+      result = start_key <= row;
+      break;
+    case KeyType::NOT_SET:
+      result = true;
+  }
+
+  // Check for end key.
+  switch (end_type) {
+    case KeyType::OPEN:
+      result = result && (row < end_key);
+      break;
+    case KeyType::CLOSED:
+      result = result && (row <= end_key);
+      break;
+    case KeyType::NOT_SET:
+      break;
+  }
+
+  return result;
+}
+
+/*
+ * Two ranges are equal, if:
+ *  - both are empty.
+ *  - start keys and types are same; in case the key value is empty,
+ *    igore the equality condition on their types (This additional check is
+ *    necessary for consistent range representation,
+ *      infinite_range() == range("", "").
+ *  - 'end' values and their types are the same.
+ *
+ * For simplicity, we omit the special boundary case:
+ *   An inifinite range with start NOT_SET or set to "" (empty string)
+ *   of type CLOSED. Both should be treated equal.
+ */
+bool RowRange::operator==(RowRange const& rhs) const {
```

Comment:
Can you simplify this to:

```
  return same_start() and same_end();
```

where each one of `same_start()` and `same_end()` sort of write themselves?